File: Scraper/models/manga.py
from dataclasses import dataclass
from peewee import *


class MangaModel(Model):
    id = CharField(column_name="ID", primary_key=True)
    authors = CharField(column_name="AUTHORS")
    categories = CharField(column_name="CATEGORIES")
    description = TextField(column_name="DESCRIPTION")
    is_ended = BooleanField(column_name="IS_ENDED")
    latest = CharField(column_name="LATEST")
    thumbnail = CharField(column_name="THUMBNAIL")
    title = CharField(column_name="TITLE")
    update_time = IntegerField(column_name="UPDATE_TIME")
    extra_data = CharField(column_name="EXTRA_DATA")

    class Meta:
        table_name = "MANGA"


class ChapterModel(Model):
    manga = ForeignKeyField(column_name="MANGA_ID", model=MangaModel)
    id = CharField(column_name="ID")
    idx = IntegerField(column_name="IDX")
    is_extra = BooleanField(column_name="IS_EXTRA")
    title = CharField(column_name="TITLE")
    urls = TextField(column_name="URLS", null=True)

    class Meta:
        table_name = "CHAPTER"
        primary_key = CompositeKey("manga", "id")


@dataclass
class Chapter:
    title: str
    id: str


@dataclass
class Chapters:
    serial: list[Chapter]
    extra: list[Chapter]
    extra_data: str


@dataclass
class PreviewManga:
    id: str
    latest: str


@dataclass
class Manga:
    categories_list = [
        "Passionate",
        "Love",
        "Campus",
        "Yuri",
        "BL",
        "Adventure",
        "Harem",
        "SciFi",
        "War",
        "Suspense",
        "Speculation",
        "Funny",
        "Fantasy",
        "Magic",
        "Horror",
        "Ghosts",
        "History",
        "FanFi",
        "Sports",
        "Hentai",
        "Mecha",
        "Restricted",
        "Otokonoko",
    ]
    id: str
    thumbnail: str
    title: str
    description: str
    is_ended: bool
    authors: list
    categories: tuple
    chapters: Chapters
    latest: str
    update_time: int

    def to_model(self):
        model = MangaModel()

        model.id = self.id
        model.title = self.title
        model.thumbnail = self.thumbnail
        model.description = self.description
        model.is_ended = self.is_ended
        model.latest = self.latest
        model.update_time = self.update_time
        model.extra_data = self.chapters.extra_data
        model.authors = "|".join(self.authors)
        model.categories = "|".join(self.categories)

        return model

    def to_chapter_models(self):
        models = []

        for idx, i in enumerate(reversed(self.chapters.serial)):
            model = ChapterModel()
            model.title = i.title
            model.id = i.id
            model.idx = idx
            model.is_extra = False
            model.manga = self.id
            models.append(model)

        for idx, i in enumerate(reversed(self.chapters.extra)):
            model = ChapterModel()
            model.title = i.title
            model.id = i.id
            model.idx = idx
            model.is_extra = True
            model.manga = self.id
            models.append(model)

        return models

File: Scraper/src/models/manga.ts
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

enum Genre {
  HotBlooded = "HotBlooded",
  Romance = "Romance",
  Campus = "Campus",
  Yuri = "Yuri",
  BL = "BL",
  Adventure = "Adventure",
  Harem = "Harem",
  SciFi = "SciFi",
  War = "War",
  Suspense = "Suspense",
  Speculation = "Speculation",
  Funny = "Funny",
  Fantasy = "Fantasy",
  Magic = "Magic",
  Horror = "Horror",
  Ghosts = "Ghosts",
  Historical = "Historical",
  FanFi = "FanFi",
  Sports = "Sports",
  Hentai = "Hentai",
  Mecha = "Mecha",
  Restricted = "Restricted",
  Otokonoko = "Otokonoko",
}

interface Chapter {
  id: string;
  title: string;
  urls?: Array<string>;
}

interface Chapters {
  extra: Array<Chapter>;
  serial: Array<Chapter>;
  extraData: string;
}

class Manga {
  constructor(
    public id: string,
    public title: string,
    public authors: Array<string>,
    public genres: Array<Genre>,
    public description: string,
    public thumbnail: string,
    public isEnded: boolean,
    public latest: string,
    public updateTime: number,
    public chapters: Chapters
  ) {}

  async save() {
    const manga: any = {};

    for (const [key, value] of Object.entries(this)) {
      switch (key) {
        case "authors":
        case "genres":
          manga[key] = value.join("|");
          break;
        case "isEnded":
        case "updateTime":
          const match = key.match(/([A-Z])/)!;
          manga[key.replace(/([A-Z])/, `_${match[1].toLowerCase()}`)] = value;
          break;
        case "chapters":
          const chapters = value as Chapters;

          manga["extra_data"] = chapters.extraData;
          manga["chapters"] = [
            ...chapters.serial.map((v, i) => ({
              is_extra: false,
              manga_id: this.id,
              idx: i,
              ...v,
            })),
            ...chapters.extra.map((v, i) => ({
              is_extra: true,
              manga_id: this.id,
              idx: i,
              ...v,
            })),
          ];
          break;
        default:
          manga[key] = value;
          break;
      }
    }

    const chapters = manga.chapters;
    delete manga.chapters;

    await prisma.manga.upsert({
      where: { id: this.id },
      create: manga,
      update: manga,
    });

    // remove deleted chapters
    await prisma.chapter.deleteMany({
      where: {
        manga_id: this.id,
        id: { notIn: chapters.map((v: any) => v.id) },
      },
    });

    // get existing chapters
    const existingIds = (
      await prisma.chapter.findMany({
        where: { manga_id: this.id },
        select: { id: true },
      })
    ).map((v) => v.id);

    // save the new chapters
    await prisma.chapter.createMany({
      data: chapters.filter((r: any) => !existingIds.includes(r.id)),
    });
  }
}

export default Manga;
export { Chapter, Chapters, Genre };
